src/utils/multipart.js

```diff
@@ -0,0 +1,113 @@
+'use strict'
+
+const Transform = require('stream').Transform
+
+const PADDING = '--'
+const NEW_LINE = '\r\n'
+const NEW_LINE_BUFFER = Buffer.from(NEW_LINE)
+
+class Multipart extends Transform {
+  constructor (options) {
+    super(Object.assign({}, options, { objectMode: true }))
+
+    this._boundary = this._generateBoundary()
+    this._files = []
+    this._draining = false
+  }
+
+  _flush () {
+    this.push(Buffer.from(PADDING + this._boundary + PADDING + NEW_LINE))
+    this.push(null)
+  }
+
+  _generateBoundary () {
+    var boundary = '--------------------------'
+    for (var i = 0; i < 24; i++) {
+      boundary += Math.floor(Math.random() * 10).toString(16)
+    }
+
+    return boundary
+  }
+
+  _transform (file, encoding, callback) {
+    if (Buffer.isBuffer(file)) {
+      this.push(file)
+      return callback() // early
+    }
+    // not a buffer, must be a file
+    this._files.push(file)
+    this._maybeDrain(callback)
+  }
+
+  _maybeDrain (callback) {
+    if (!this._draining) {
+      if (this._files.length) {
+        this._draining = true
+        const file = this._files.shift()
+        this._pushFile(file, (err) => {
+          this._draining = false
+          if (err) {
+            this.emit('error', err)
+          } else {
+            this._maybeDrain(callback)
+          }
+        })
+      } else {
+        this.emit('drained all files')
+        callback()
+      }
+    } else {
+      this.once('drained all files', callback)
+    }
+  }
+
+  _pushFile (file, callback) {
+    const leading = this._leading(file.headers || {})
+
+    this.push(leading)
+
+    let content = file.content || Buffer.alloc(0)
+
+    if (Buffer.isBuffer(content)) {
+      this.push(content)
+      this.push(NEW_LINE_BUFFER)
+      return callback() // early
+    }
+
+    // From now on we assume content is a stream
+
+    content.once('error', this.emit.bind(this, 'error'))
+
+    content.once('end', () => {
+      this.push(NEW_LINE_BUFFER)
+      callback()
+
+      // TODO: backpressure!!! wait once self is drained so we can proceed
+      // This does not work
+      // this.once('drain', () => {
+      //   callback()
+      // })
```

Comment:
No needed anymore?

Reply:
We do back-pressure per file item, but once inside a file we don't. So it's optimised for a lot of files, but not for large files. If that comes to be a problem, I think we should open a different issue to tackle it..

Reply:
Mind writing this note as the comment as well?

Reply:
Decided to fix individual file back-pressure in this PR: https://github.com/ipfs/js-ipfs-api/pull/629/commits/782beb9a628b5d9bde047400d1d44f6d57b17ef8

Reply:
Had to limit it to Node only though, because browser HTTP request does not stream..
7403afc368c2992a4e19f4ac14b0815fbcb3c045